core/src/main/scala/org/apache/spark/scheduler/SchedulerBackend.scala

```diff
@@ -92,4 +93,16 @@ private[spark] trait SchedulerBackend {
    */
   def maxNumConcurrentTasks(rp: ResourceProfile): Int
 
+  /**
+   * Get the list of host locations for push based shuffle
+   *
+   * Currently push based shuffle is disabled for both stage retry and stage reuse cases
```

Comment:
This looks a little inconsistent with https://github.com/apache/spark/pull/30164/files#diff-85de35b2e85646ed499c545a3be1cd3ffd525a88aae835a9c621f877eebadcb6R1314 (`This way, even if this stage gets retried, it would still be merging blocks using the same set...`).

Reply:
So, both are true as of today. Did I understand correctly?

Reply:
Both are true.
`getShufflePushMergerLocations` will be invoked only once per `ShuffleDependency`.
Thus retried stages will get the same merger locations.
In #30062, the way we implemented the block push handling logic would ignore blocks received after shuffle finalization.

https://github.com/apache/spark/blob/dd32f45d2058d00293330c01d3d9f53ecdbc036c/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java#L132

So, blocks pushed from the retry stage will be ignored, and this is what's stated in this comment about `push based shuffle is disabled for both stage retry and stage reuse cases`.
Ignoring blocks pushed from the retry stage is reasonable, since the block data from these retried tasks most likely have already been merged.
Making sure the retried stage use the same merger location is critical to ensure we don't run into data duplication issues.

The only exception is for indeterministic stage retry, which we have created SPARK-32923 for it.